### PR TITLE
Email Integration for API3 Explorer

### PR DESCRIPTION
**Pull request by aquarat**
This PR implements an email handler for the API3 Explorer.

Associated interim code in the explorer is available here: https://github.com/api3dao/api3-explorer/blob/9b6fc24475d0eddbf8dad8a87da339da58fed2d3/src/pages/order/order.tsx#L185

This PR also adds some tests to other modules in the project.

### CODE REVIEW

File: src/validate-and-send.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
+import axios from 'axios';
+import { APIGatewayProxyHandler } from 'aws-lambda';
+import { z } from 'zod';
+import { go } from '@api3/airnode-utilities';
+import { Client } from '@sendgrid/client';
+// There doesn't appear to be a way around require here
+import sgMail = require('@sendgrid/mail');
```
[the remark this answers is not in the document]
Comment (aquarat) [reply src/validate-and-send.ts:7]: No it doesn't. I'm not sure why. I asked the Protofire guy to try and not use require but I see now importing doesn't work for some reason.

File: src/validate-and-send.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import { getGlobalConfig } from './utils';
+
+sgMail.setClient(new Client());
+
+const config = getGlobalConfig();
+
+// The contact form POST payload schema
+// We don't want to constrain this schema too much as it may change.
```
[the remark this answers is not in the document]
Comment (aquarat) [reply src/validate-and-send.ts:15]: We're very constrained for resources currently and they've already altered the schema.
They also need to add some additional data to this.
We can probably just keep this PR open until the payload schema has stabilised.

File: src/validate-and-send.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  SENDGRID_API_KEY: z.string(),
+});
+
+/**
+ * Validates the ReCaptcha token in a POST payload body and if valid proceeds to send an email with the contact form
+ * data.
+ *
+ * @param event an AWS API Gateway Event
+ */
+export const validateAndSendEmail: APIGatewayProxyHandler = async (event): Promise<any> => {
```
[the remark this answers is not in the document]
Comment (aquarat) [reply src/validate-and-send.ts:42]: Of course :see_no_evil:

File: src/validate-and-send.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    return {
+      statusCode: 400,
+      headers: config.headers,
+      body: 'Invalid post body',
+    };
+  }
+
+  const contactForm = safeBody.data;
+
+  const [err, isValidCaptcha] = await go(async () => await validateToken(contactForm.token), { timeoutMs: 5_000 });
```
[the remark this answers is not in the document]
Comment (aquarat) [reply src/validate-and-send.ts:63]: Thanks, I will avoid redundant `async`/`await` in future :)

**Comment by aquarat** [comment]
I'm just pasting the revised data structure here myself/others to aid finalisation of the validation schema:
```
Order: {
  "chainId": "137",
  "chainName": "polygon",
  "items": [
    {
      "beaconName": "AAVE/USD",
      "beaconId": "0x6b1d92efa7d87e2185595819828f7f884ee4ff09fe468cb2100f86649aa09ebe",
      "coverage": 500000,
      "months": 3,
      "fee": 10000
    }
  ]
}
```

**Comment by aquarat** [comment]
I've opened a PR on the explorer with Zod schema additions. Once that's accepted I'll introduce that schema to this PR.

https://github.com/api3dao/api3-explorer/pull/52

**Comment by aquarat** [comment]
I'm really struggling to test the validate-and-send.ts file because the sgMail client causes jest to throw an error:
```
  ● Test suite failed to run

    Configuration error:
    
    Could not locate module ./src/client mapped as:
    /home/aquarat/Projects/beacons-api/src/$1.
    
    Please check your configuration for these entries:
    {
      "moduleNameMapper": {
        "/src\/(.*)/": "/home/aquarat/Projects/beacons-api/src/$1"
      },
      "resolver": undefined
    }

      at createNoMappedModuleFoundError (node_modules/jest-resolve/build/resolver.js:579:17)
      at Object.<anonymous> (node_modules/@sendgrid/client/index.js:3:16)

```
The email handler is supposed to be a stop-gap until Monday.com integration is done and that is a stop-gap until a full payment flow is implemented, so I'm not sure it's worth spending more time on this.

File: serverless.example.yml [hunk 5]
```diff
@@ -49,6 +49,17 @@ functions:
       - http:
           path: beacons/on_chain_value
           method: get
+          # Functions often require an associated CORS OPTIONS method enabled on the API gateway to satisfy Chrome pre-flight
+          # requests. An OPTIONS method is generally not required for GET requests but is always required by Chrome for
```
Comment (Siegrift) [inline serverless.example.yml:53]: This comment is not true. CORS is not required for certain POST requests. I'd just comment this:

```
Enable CORS for every endpoint. Based on https://www.serverless.com/blog/cors-api-gateway-survival-guide/
```

File: serverless.example.yml [hunk 6]
```diff
@@ -61,19 +72,25 @@ functions:
   last_transactions:
     handler: src/last-transactions.lastTransactions
     timeout: 30
-    memorySize: 256
+    memorySize: 128
```
Comment (Siegrift) [inline serverless.example.yml:75]: I see you decreased the `memorySize`, `ttlInSeconds` and `schedule`. Are the new values expected?

File: serverless.example.yml [hunk 7]
```diff
@@ -101,9 +126,30 @@ functions:
...
           method: get
+          cors:
+            headers: '*'
+            origin: 'http://domain-one.com,https://domain-two.com'
+            methods:
+              - OPTIONS
+              - GET
           caching:
-            enabled: true
+            enabled: false
```
Comment (Siegrift) [inline serverless.example.yml:136]: Why disable caching here?

File: src/dapi-names.test.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
+import { Client } from 'pg';
+import { ethers } from 'ethers';
+import { dapiNameQueryTemplate, getDataFeedIdFromDapiName } from './dapi-names';
+
+describe('handles dapi name query', () => {
+  it('responds to a valid query', async () => {
+    const mockQueryImplementation = jest.fn().mockImplementation((_query: string, _parameters: any[]) => ({
```
Comment (Siegrift) [inline src/dapi-names.test.ts:7]: You can just omit the parameters here and in the mock below.

File: src/dapi-names.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
+import { Client } from 'pg';
+import { ethers } from 'ethers';
+import { go } from '@api3/promise-utils';
+import { goQueryConfig } from './constants';
+
+export const dapiNameQueryTemplate = `
```
Comment (Siegrift) [inline src/dapi-names.ts:6]: How to test this locally? 

It would be nice to have some test DB because now it's not possible to test/deploy this on it's own. However, this should be done in a separate PR.

File: src/last-transactions.ts [hunk 10]
```diff
@@ -1,24 +1,32 @@
...
 
 export const dapiNameSchema = z.string();
 export const evmBeaconIdSchema = z.string().regex(/^0x[a-fA-F0-9]{64}$/);
-export const evmAddressSchema = z.string().regex(/^0x[a-fA-F0-9]{40}$/);
 export const chainIdSchema = z.string().regex(/^\d+$/);
 
-export const DEFAULT_TRANSACTION_COUNT = 5;
-export const DEFAULT_FROM_BLOCK_SIZE = 2000;
-export const MAX_TRANSACTIONS_FRESHNESS = 60_000;
+export const DEFAULT_TRANSACTION_COUNT = 40;
```
Comment (Siegrift) [inline src/last-transactions.ts:13]: Nit: This one seems not overridable so it could be `TRANSACTION_COUNT` instead.

File: src/last-transactions.ts [hunk 11]
```diff
@@ -75,26 +83,10 @@ export const lastTransactions: APIGatewayProxyHandler = async (event): Promise<a
...
-    event_data -> 'topics' as "topics",
-    event_data -> 'transactionHash' as "transactionHash"
-    FROM dapi_events 
-    WHERE chain = $1 AND 
-    event_data->'parsedLog'->'args'->> 0 = $2
-    ORDER by block DESC LIMIT $3;
-  `,
-      [queryChainId, queryBeaconId, queryTransactionCountLimit]
-    );
+    db.query(lastTransactionsQueryTemplate, [queryChainId, queryBeaconId, queryTransactionCountLimit]);
```
Comment (Siegrift) [inline src/last-transactions.ts:89]: It's better to inline the `lastTransactionsQueryTemplate` since it could go out of sync with the arguments (since it's defined way above the usage).

File: src/last_transactions.test.ts [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+    const result = await lastTransactions(
+      {
+        queryStringParameters,
+      } as unknown as APIGatewayProxyEvent,
+      {} as unknown as Context,
+      () => {}
+    );
+    expect(initDb).toHaveBeenCalledTimes(1);
+    expect(mockQueryImplementation).toHaveBeenCalledTimes(1);
+    expect(mockQueryImplementation).toHaveBeenCalledWith(lastTransactionsQueryTemplate, [
```
Comment (Siegrift) [inline src/last_transactions.test.ts:36]: I'd say this is too much detail and would remove that assertion (it also prevents inlining the query which feels more important).

File: src/last_transactions.test.ts [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+      queryStringParameters.chainId,
+      queryStringParameters.beaconId,
+      DEFAULT_TRANSACTION_COUNT,
+    ]);
+
+    expect(result).toBeDefined();
```
Comment (Siegrift) [inline src/last_transactions.test.ts:42]: This assertion is not necessary and you can use `!` for the ones below.

(Similarly for other tests).

File: src/validate-and-send.ts [hunk 14]
```diff
@@ -0,0 +1,152 @@
+import axios from 'axios';
+import { go } from '@api3/promise-utils';
+import { APIGatewayProxyHandler } from 'aws-lambda';
+import { z } from 'zod';
+import { Client } from '@sendgrid/client';
+// There doesn't appear to be a way around `require` here
```
Comment (Siegrift) [inline src/validate-and-send.ts:6]: Nit: 
```suggestion
// There doesn't seem to be a way around `require` here
```

File: src/last_transactions.test.ts [hunk 15]
```diff
@@ -0,0 +1,239 @@
+import * as fs from 'fs';
```
Comment (Siegrift) [inline src/last_transactions.test.ts:1]: The filename could be called `last-transaction.test.ts` for consistency.

File: src/validate-and-send.ts [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+export const validateAndSendEmail: APIGatewayProxyHandler = async (event) => {
+  if (!event.body) {
+    return {
+      statusCode: 400,
+      headers: config.headers,
+      body: 'Invalid post body',
+    };
+  }
+
+  if (process.env.DEBUG) {
```
Comment (Siegrift) [inline src/validate-and-send.ts:39]: This if is already handled in the function.

File: test/fixtures/last-transactions-example.json [hunk 17]
```diff
@@ -0,0 +1,5602 @@
+[
```
Comment (Siegrift) [inline test/fixtures/last-transactions-example.json:1]: Nit: Could we shorten this fixture a bit?

**Review by Siegrift — COMMENTED** [review]
I just did a first pass over the code, but there doesn't seem to be an easy way how to test this locally.

File: serverless.example.yml (hunk 6 above)
Comment (aquarat) [reply serverless.example.yml:75]: `ttlSeconds` is subjective :shrug: it defines the cache timeout. The longer the cache timeout the less these functions will cost us.
`schedule` is also arguably subjective, but they want a high resolution history of prices, so I figured one minute was ideal.
The functions work fine with 128 MB of memory and it naturally costs us less. These functions are predominately I/O bound.

File: serverless.example.yml (hunk 5 above)
Comment (aquarat) [reply serverless.example.yml:53]: Ah I see, yes, it's more complex than just the POST method. Thanks for correcting me.

File: serverless.example.yml (hunk 7 above)
Comment (aquarat) [reply serverless.example.yml:136]: This is a mistake, thank you, I'll fix it.

File: src/dapi-names.ts (hunk 9 above)
Comment (aquarat) [reply src/dapi-names.ts:6]: Yes, ideally one should probably start a test DB in the CI system maybe.

Comment (Siegrift) [reply src/dapi-names.ts:6]: Localhost would be enough imo. Currently, you are the only one who is able to conveniently run this :/

Comment (aquarat) [reply src/dapi-names.ts:6]: I plan on storing the latest deployment files in Keybase, which anyone with access will be able to use to redeploy and/or test (with production data).
But I totally agree with you that we need a test DB :)

**Comment by aquarat** [comment]
@Siegrift I believe I have resolved the main issues?
Thanks for reviewing :)

File: README.md [hunk 18]
```diff
@@ -27,8 +27,24 @@ cp serverless.example.yml serverless.yml
...
 ```shell
-rm -rf .build .serverless;
+rm -rf .build .serverless; # Sometimes serverless caches build artifacts that can confound testing
+yarn sls deploy
+```
+
+To test locally:
+
+```shell
+rm -rf .build .serverless; # Sometimes serverless caches build artifacts that can confound testing
```
Comment (Siegrift) [inline README.md:47]: Btw. these instructions will not work, since the DB is missing, right?

**Review by Siegrift — APPROVED** [review]
👍 LGTM

File: README.md (hunk 18 above)
Comment (aquarat) [reply README.md:47]: The instructions assume (😬) that you have the correct database credentials set locally.

Comment (aquarat) [reply README.md:47]: I will update these instructions accordingly 👍
